libcudacxx/include/cuda/std/detail/libcxx/include/optional

```diff
@@ -777,172 +713,151 @@ public:
+                ( _LIBCUDACXX_TRAIT(is_constructible, _Tp, _Up)) _LIBCUDACXX_AND
+                (!_LIBCUDACXX_TRAIT(is_convertible, _Up, _Tp)))
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr
+    explicit optional(_Up&& __v) : __base(in_place, _CUDA_VSTD::forward<_Up>(__v)) {}
+
+    _LIBCUDACXX_TEMPLATE(class _Up)
+      (requires (!_LIBCUDACXX_TRAIT(is_same, _Up, _Tp)) _LIBCUDACXX_AND
+                ( _LIBCUDACXX_TRAIT(is_constructible, _Tp, _Up const&)) _LIBCUDACXX_AND
+                ( _LIBCUDACXX_TRAIT(is_convertible, _Up const&, _Tp)) _LIBCUDACXX_AND
+                (!__check_constructible_from_opt<_Up>::value))
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr
+    optional(const optional<_Up>& __v) {
         this->__construct_from(__v);
     }
-    template <class _Up, __enable_if_t<
-        _CheckOptionalLikeCtor<_Up, _Up const&>::template __enable_explicit<_Up>::__value
-    , int> = 0>
-    _LIBCUDACXX_INLINE_VISIBILITY
-    constexpr explicit optional(const optional<_Up>& __v)
-    {
+
+    _LIBCUDACXX_TEMPLATE(class _Up)
+      (requires (!_LIBCUDACXX_TRAIT(is_same, _Up, _Tp)) _LIBCUDACXX_AND
+                ( _LIBCUDACXX_TRAIT(is_constructible, _Tp, _Up const&)) _LIBCUDACXX_AND
+                (!_LIBCUDACXX_TRAIT(is_convertible, _Up const&, _Tp)) _LIBCUDACXX_AND
+                (!__check_constructible_from_opt<_Up>::value))
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr
+    explicit optional(const optional<_Up>& __v) {
         this->__construct_from(__v);
     }
 
-    // LWG2756: conditionally explicit conversion from optional<_Up>&&
-    template <class _Up, __enable_if_t<
-        _CheckOptionalLikeCtor<_Up, _Up &&>::template __enable_implicit<_Up>::__value
-    , int> = 0>
-    _LIBCUDACXX_INLINE_VISIBILITY
-    constexpr optional(optional<_Up>&& __v)
-    {
+    _LIBCUDACXX_TEMPLATE(class _Up)
+      (requires (!_LIBCUDACXX_TRAIT(is_same, _Up, _Tp)) _LIBCUDACXX_AND
+                ( _LIBCUDACXX_TRAIT(is_constructible, _Tp, _Up)) _LIBCUDACXX_AND
+                ( _LIBCUDACXX_TRAIT(is_convertible, _Up, _Tp)) _LIBCUDACXX_AND
+                (!__check_constructible_from_opt<_Up>::value))
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr
+    optional(optional<_Up>&& __v) {
         this->__construct_from(_CUDA_VSTD::move(__v));
     }
-    template <class _Up, __enable_if_t<
-        _CheckOptionalLikeCtor<_Up, _Up &&>::template __enable_explicit<_Up>::__value
-    , int> = 0>
-    _LIBCUDACXX_INLINE_VISIBILITY
-    constexpr explicit optional(optional<_Up>&& __v)
-    {
+
+    _LIBCUDACXX_TEMPLATE(class _Up)
+      (requires (!_LIBCUDACXX_TRAIT(is_same, _Up, _Tp)) _LIBCUDACXX_AND
+                ( _LIBCUDACXX_TRAIT(is_constructible, _Tp, _Up)) _LIBCUDACXX_AND
+                (!_LIBCUDACXX_TRAIT(is_convertible, _Up, _Tp)) _LIBCUDACXX_AND
+                (!__check_constructible_from_opt<_Up>::value))
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr
+    explicit optional(optional<_Up>&& __v) {
```

Comment:
I don't like this approach. It muddles the purpose of those conditions, forces me to check them in all places to verify they are the same, and generates a lot more code than was there before. If this change is absolutely necessary, I would like you to find a way to give names to these sets of conditions, instead of repeating them over and over again. Same comment on the assignment operators.

I thought I've ran this with 2017, how is it choking on this?

Reply:
The issue is that MSVC 2017 is falling over its feet when confronted with
```
explicit optional(const optional<_Up>& __v)
```

Because it confuses that with the copy constructor. The same applies for any variadic constructor, where we have the same issue. But that applies more to `tuple` than `optional`

I was also not that happy with the amount of code. I will have a look whether I can pull it into static variables.

Reply:
I have reworked the approach to be much more expressive. I hope that is about what you had in mind